### PR TITLE
Wordbound blanks in Recursive Transfer

### PR DESCRIPTION
**Pull request by khannatanmai**
- Wordbound blanks are read as part of LUs as a new side->wblank.
- Wblanks reorder with the LUs in transfer based on where the lemma is clipped from.
- Works even if lemma is clipped into a variable and the variable is later added in the output.
- No regression. Stream without wordbound blanks work as-is.
- Normal blanks don't move around while wordbound blanks move around.
- Tests added.
- When MLUs are formed the blanks are merged.
- MLU tests added.

### CODE REVIEW

File: src/rtx_processor.cc [hunk 1]
```diff
@@ -319,6 +394,14 @@ RTXProcessor::applyRule(const wstring& rule)
   const wchar_t* rule_data = rule.data();
   for(unsigned int i = 0, rule_size = rule.size(); i < rule_size; i++)
   {
+    /*
```
Comment (mr-martian) [inline src/rtx_processor.cc:397]: extraneous debug comment

File: tests/WblankVariableTRX.trx [hunk 2]
```diff
@@ -0,0 +1,95 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (mr-martian) [inline tests/WblankVariableTRX.trx:1]: The rtx equivalent is probably
```
gender = m f mf;
number = sg pl sp;

n: _.gender.number;
adj: _.gender.number;
vblex: _.number;
NP: _;
V: _;

NP -> adj %n [$%number=2.number/tl, $%lemvar=1.lemh/tl] { 2 _1 1[lemh=$%lemvar, number=2.number/sl] } ;

V -> %vblex [$%number=(if ($%number = "") sg else $%number), $%lemvar=1.lemh/tl] { 1[lemh=$%lemvar, number=$%number] } ;
```

Comment (khannatanmai) [reply tests/WblankVariableTRX.trx:1]: Syntax Error hmm:
```
$ rtx-comp tests/WblankVariable.rtx test.bin
Syntax error on line 10 of tests/WblankVariable.rtx: expected '/', found '$'
NP -> adj %n [$%number=2.number/tl, $%lemvar=1.lemh/tl] { 2 _1 1[lemh=$%lemvar, number=2.number/sl] } ;
                                   ^^^
```

File: src/rtx_processor.cc (hunk 1 above)
Comment (khannatanmai) [reply src/rtx_processor.cc:397]: will remove in last commit

File: tests/WblankVariableTRX.trx (hunk 2 above)
Comment (mr-martian) [reply tests/WblankVariableTRX.trx:1]: syntax issue fixed in 9982536

**Comment by khannatanmai** [comment]
> Only thing that stands out from a C++ point-of-view is the C-style casts, but those are unfortunately common in Apertium code.

TinoDidriksen where's the casts? No time like the present to fix them

**Comment by khannatanmai** [comment]
> `Chunk::conjoin()` should also merge blanks.

@mr-martian, @TinoDidriksen  so I ran across a problem, and there's several fixes, just wondering what's the best one: 
chunk.h has conjoin(), which needs to call the function combineWblanks, which is declared in rtx_processor.h. The function is also needed in rtx_processor.cc. Now, rtx_processor.h includes chunk.h, what I want is that both chunk.h and rtx_processor.cc can access combineWblanks(). Naturally, I tried moving combineWblanks to chunk.h so that both can access it, but it leads to some linker errors that can be fixed by making the function inline. Here are the possible solutions:

1. Leave it as an inline function. It is a big function, so that's not ideal.
2. Keep the function in rtx_processor.h and include rtx_processor.h in chunk.h as well. I'm not sure if this cyclical inclusion causes any problems, but yeah just proposing.
3. Instead of combining wblanks inside conjoin(), I do it inside rtx_processor.cc, which is what I'm doing right now. This works, but the issue is that if conjoin is called elsewhere at some point, the wblank combining would need to happen again, plus philosophically the combining of wblanks should happen inside conjoin().
4. I write the same function in chunk.h and use that. A little duplication of code but all should be well.

Or, if there's just some way to fix the linker error then even better. Here's what I see after just moving the function to chunk.h:
```
g++  -Wall -Wextra -O2 -DSIZET_NOT_CSTDINT=1 -std=c++2a -lxml2  -L/Users/khannatanmai/Documents/GSoC/repo/main/lib -L/opt/local/lib -lapertium3 -llttoolbox3 -lxml2 -lpcre -L/opt/local/lib -lxml2  -o rtx-proc rtx_proc.o rtx_processor.o apertium_re.o  -lxml2  -L/Users/khannatanmai/Documents/GSoC/repo/main/lib -L/opt/local/lib -lapertium3 -llttoolbox3 -lxml2 -lpcre -L/opt/local/lib -lxml2
duplicate symbol 'combineWblanks(std::__1::basic_string<wchar_t, std::__1::char_traits<wchar_t>, std::__1::allocator<wchar_t> >, std::__1::basic_string<wchar_t, std::__1::char_traits<wchar_t>, std::__1::allocator<wchar_t> >)' in:
    rtx_proc.o
    rtx_processor.o
ld: 1 duplicate symbol for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [rtx-proc] Error 1
make[1]: *** [all] Error 2
make: *** [all-recursive] Error 1
```

**Comment by TinoDidriksen** [comment]
I would split chunk.h into chunk.h / chunk.cc. Only need to move `conjoin()` into the .cc for now.

**Comment by khannatanmai** [comment]
> I would split chunk.h into chunk.h / chunk.cc. Only need to move `conjoin()` into the .cc for now.

Can do. @mr-martian ?

**Comment by khannatanmai** [comment]
Also, @mr-martian I just added wblank printing in the treePlain function. Which modes should display the wblanks when printed?

**Comment by mr-martian** [comment]
`writeTreePlain()` deals with everything that should print blanks.

`chunk.h` probably should have gotten a `.cc` a long time ago. If you don't do it now, I'll probably do it in a week or two.

File: src/chunk.h [hunk 3]
```diff
@@ -43,7 +46,7 @@ class Chunk
   Chunk(wstring src, wstring dest, wstring cor, wstring wbl)
   : source(src), target(dest), coref(cor), wblank(wbl), isBlank(false), isJoiner(false), rule(-1)
   {}
-  Chunk(wstring dest, vector<Chunk*>& children, int r = -1)
+  Chunk(wstring dest, vector<Chunk*>& children, int r)
```
Comment (mr-martian) [inline src/chunk.h:49]: Why was this changed?

File: src/chunk.cc [hunk 4]
```diff
@@ -0,0 +1,621 @@
...
+      lemq_loc++;
+      continue;
+    }
+    else if(target[lemq_loc] == L'#')
+    {
+      break;
+    }
+  }
+  target.insert(lemq_loc, L"+" + other->target);
+  wblank = combineWblanks(other->wblank, wblank);
```
Comment (mr-martian) [inline src/chunk.cc:296]: Seems like maybe this should be the other order?

File: src/chunk.h (hunk 3 above)
Comment (khannatanmai) [reply src/chunk.h:49]: It gives an error to have a default value in the declaration. The default value is there in the cc file so nothing really changes.

Comment (mr-martian) [reply src/chunk.h:49]: I don't see the initializers in the .cc

File: src/chunk.cc (hunk 4 above)
Comment (mr-martian) [reply src/chunk.cc:296]: as in `wblank, other->wblank`

Comment (khannatanmai) [reply src/chunk.cc:296]: Yeah it seems like it but apparently other is the first one in order and the current chunk is the second one.

Comment (khannatanmai) [reply src/chunk.cc:296]: See the conjoined LU tests

Comment (mr-martian) [reply src/chunk.cc:296]: weird, but ok

File: src/chunk.h (hunk 3 above)
Comment (khannatanmai) [reply src/chunk.h]: Right, I moved it earlier, but forgot to add it again. Added

**Comment by khannatanmai** [comment]
Alright, looks ready to merge to me.
